Mathlib/NumberTheory/ModularForms/JacobiTheta/Bounds.lean

```diff
@@ -0,0 +1,271 @@
+      filter_upwards [eventually_gt_atTop 0] with t ht
+      rw [norm_of_nonneg (exp_pos _).le, exp_le_exp]
+      nlinarith [pi_pos]
+    · simp_rw [mul_div_assoc, ← neg_mul]
+      apply IsBigO.const_mul_left
+      simpa only [mul_one] using (isBigO_refl _ _).mul aux
+
+end k_eq_one
+
+end nat
+
+section int
+
+/-- Summand in the sum to be bounded (`ℤ` version). -/
+def f_int (k : ℕ) (a t : ℝ) (n : ℤ) : ℝ := |n + a| ^ k * exp (-π * (n + a) ^ 2 * t)
+
+lemma f_int_ofNat (k : ℕ) {a : ℝ} (ha : 0 ≤ a) (t : ℝ) (n : ℕ) :
+    f_int k a t (Int.ofNat n) = f_nat k a t n := by
+  rw [f_int, f_nat, Int.ofNat_eq_coe, Int.cast_natCast, abs_of_nonneg (by positivity)]
+
+lemma f_int_negSucc (k : ℕ) {a : ℝ} (ha : a ≤ 1) (t : ℝ) (n : ℕ) :
+    f_int k a t (Int.negSucc n) = f_nat k (1 - a) t n := by
+  have : (Int.negSucc n) + a = -(n + (1 - a)) := by { push_cast; ring }
+  rw [f_int, f_nat, this, abs_neg, neg_sq, abs_of_nonneg (by linarith)]
+
+lemma summable_f_int (k : ℕ) (a : ℝ) {t : ℝ} (ht : 0 < t) : Summable (f_int k a t) := by
+  apply Summable.of_norm
+  have := (HasSum.int_rec (summable_f_nat k a ht).hasSum
+    (summable_f_nat k (1 - a) ht).hasSum).summable.norm
+  convert this using 2 with n
+  rcases n with m | m
+  · simp only [f_int, f_nat, Int.ofNat_eq_coe, Int.cast_natCast, norm_mul, norm_eq_abs, abs_pow,
+      abs_abs]
+  · simp only [f_int, f_nat, Int.cast_negSucc, norm_mul, norm_eq_abs, abs_pow, abs_abs,
+      (by { push_cast; ring } : -↑(m + 1) + a = -(m + (1 - a))), abs_neg, neg_sq]
+
+lemma _root_.UnitAddCircle.eq_coe_Ico (a : UnitAddCircle) : ∃ b : ℝ, b ∈ Ico 0 1 ∧ ↑b = a := by
+  let b := (QuotientAddGroup.equivIcoMod zero_lt_one 0) a
+  exact ⟨b.1, by simpa only [zero_add] using b.2,
+    (QuotientAddGroup.equivIcoMod zero_lt_one 0).symm_apply_apply a⟩
+
+/-- The sum to be bounded (`ℤ` version). -/
+def F_int (k : ℕ) (a : UnitAddCircle) (t : ℝ) : ℝ :=
+  (show Function.Periodic (fun b ↦ ∑' (n : ℤ), f_int k b t n) 1 by
+    intro b
+    simp_rw [← (Equiv.addRight (1 : ℤ)).tsum_eq (f := fun n ↦ f_int k b t n)]
+    simp only [f_int, ← add_assoc, add_comm, Equiv.coe_addRight, Int.cast_add, Int.cast_one]
+    ).lift a
+
+lemma F_int_eq_of_mem_Icc (k : ℕ) {a : ℝ} (ha : a ∈ Icc 0 1) {t : ℝ} (ht : 0 < t) :
+    F_int k a t = (F_nat k a t) + (F_nat k (1 - a) t) := by
+  simp only [F_int, F_nat, Function.Periodic.lift_coe]
+  convert ((summable_f_nat k a ht).hasSum.int_rec (summable_f_nat k (1 - a) ht).hasSum).tsum_eq
+    using 3 with n
+  rcases n with m | m
+  · rw [f_int_ofNat _ ha.1]
+  · rw [f_int_negSucc _ ha.2]
+
+lemma isBigO_atTop_F_int_zero_sub (a : UnitAddCircle) : ∃ p, 0 < p ∧
+    (fun t ↦ F_int 0 a t - (if a = 0 then 1 else 0)) =O[atTop] fun t ↦ exp (-p * t) := by
+  obtain ⟨a, ha, rfl⟩ := a.eq_coe_Ico
+  obtain ⟨p, hp, hp'⟩ := isBigO_atTop_F_nat_zero_sub ha.1
+  obtain ⟨n, hn, hn'⟩ := isBigO_atTop_F_nat_zero_sub (sub_nonneg.mpr ha.2.le)
+  have ha' : (a : UnitAddCircle) = 0 ↔ a = 0 := by
```

Comment:
This looks like a missing API lemma for `UnitAddCircle`.

Reply:
I've moved this to the file where AddCircle and UnitAddCircle are defined.